File: 17.letterCombinationsOfAPhoneNumber/letterCombinations_dfs.java
class Solution {
    public List<String> letterCombinations(String digits) {
         List<String> res = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return res;
        }
        String[] digit2letter = new String[]{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        StringBuilder path = new StringBuilder();
        dfs(digits, digit2letter, res, path, 0);
        return res;
    }
    
    private void dfs(String digits, String[] digit2letter, List<String> res, StringBuilder path, int level) {
        if (level == digits.length()) {
            res.add(path.toString());
            return;
        }
        String letter = digit2letter[digits.charAt(level) - '0'];
        for (char c : letter.toCharArray()) {
            path.append(c);
            dfs(digits, digit2letter, res, path, level + 1);
            path.deleteCharAt(path.length() - 1);;
        }
    }
}

File: 17.letterCombinationsOfAPhoneNumber/letterCombinations_dfs.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.length() == 0) return res;
        string path;
        vector<string> digit2letters = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        dfs(digits, 0, path, res, digit2letters);
        return res;
    }
    void dfs(string digits, int level, string& path, vector<string>& res, vector<string>& digit2letters) {
        if (level == digits.length()) {
            res.push_back(path);
            return;
        }
        for (char c : digit2letters[digits[level] - '0']) {
            path.push_back(c);
            dfs(digits, level + 1, path, res, digit2letters);
            path.pop_back();
        }
    }
};

int main()
{
    Solution s;
    vector<string> ans = s.letterCombinations(string("234"));
    vector<string>::iterator it = ans.begin();
    while(it != ans.end()){
        cout << *it << " ";
        ++it;
    }
    cout << endl;
    return 0;
}
